mqtt.cpp

```diff
@@ -0,0 +1,231 @@
+}
+
+static void mqtt_disconnection_cb(struct mosquitto *mqtt_client, void *obj, int rc) {
+	DEBUG_LOGF("MQTT Disconnection: %s\n", mosquitto_strerror(rc));
+}
+
+static void mqtt_log_cb(struct mosquitto *mqtt_client, void *obj, int level, const char *message){
+	if (level != MOSQ_LOG_DEBUG )
+	DEBUG_LOGF("MQTT Log: %s (%d)\n", message, level);
+}
+
+#endif
+
+bool OSMqtt::enabled(void) { return _enabled; }
+
+// Configure the broker host and port fields and enabled/disable the interface
+// param config pointer to a string containing the JSON configuration settings in the form of "{server:"host_name|IP address", port: 1883, enabled:"0|1"}"
+void OSMqtt::setup(void) {
+	String config = os.sopt_load(SOPT_MQTT_OPTS);
+	if (config.length() == 0) {
+		_enabled = false;
+		DEBUG_LOGF("MQTT Setup: Config (None) Disabled\n");
+	} else {
+		int enable = false;
+		sscanf(config.c_str(), "\"server\":\"%[^\"]\",\"port\":\%d,\"enable\":\%d", _host, &_port, &enable);
+		_enabled = (bool)enable;
+		DEBUG_LOGF("MQTT Setup: Config (%s:%d) %s\n", _host, _port, _enabled ? "Enabled" : "Disabled");
+	}
+
+	if (mqtt_client == NULL) return;
+
+#if defined(ARDUINO)
+	if (mqtt_client->connected()) {
+		mqtt_client->disconnect();
+	}
+
+	if (_enabled) {
+		mqtt_client->setServer(_host, _port);
+		if (mqtt_client->connect(_id, NULL, NULL, "opensprinkler/availability", 0, true, "offline")) {
+			mqtt_client->publish("opensprinkler/availability", "online", true);
+		} else {
+			DEBUG_LOGF("MQTT Setup: Connect Failed (%d)\n", mqtt_client->state());
+		}
+	}
+#elif defined(OSPI)
+	mosquitto_disconnect(mqtt_client);
+
+	if (_enabled) {
+		int rc = mosquitto_connect(mqtt_client, _host, _port, MQTT_KEEPALIVE);
+		if (rc != MOSQ_ERR_SUCCESS) DEBUG_LOGF("MQTT Setup: Connect Failed (%s)\n", mosquitto_strerror(rc));
+	}
+#endif
+}
+
+void OSMqtt::publish(const char *topic, const char *payload) {
+	DEBUG_LOGF("MQTT Publish: %s %s\n", topic, payload);
+	if (mqtt_client == NULL || !_enabled || os.status.network_fails > 0) return;
+
+#if defined (ARDUINO)
+	if (!mqtt_client->publish(topic, payload)) {
+		DEBUG_LOGF("MQTT Publish: Failed (%d)\n", mqtt_client->state());
+	}
+#elif defined(OSPI)
+	int rc = mosquitto_publish(mqtt_client, NULL, topic, strlen(payload), payload, 0, false);
+	if (rc != MOSQ_ERR_SUCCESS) {
+		DEBUG_LOGF("MQTT Publish: Failed (%s)\n", mosquitto_strerror(rc));
+	}
+#endif
+}
+
+// Need to regularly call the loop function to ensure "keep alive" messages are sent to the broker and to reconnect if needed.
+void OSMqtt::loop(void) {
+	static unsigned long last_reconnect_attempt = 0;
+	unsigned long now = millis();
+
+	if (mqtt_client == NULL || !_enabled || os.status.network_fails > 0) return;
+
+#if defined (ARDUINO)
+	static int last_state = 999;
+	if (!mqtt_client->connected() && (now - last_reconnect_attempt >= (unsigned long)MQTT_RECONNECT_DELAY*1000)) {
+		if (mqtt_client->connect(_id, NULL, NULL, "opensprinkler/availability", 0, true, "offline"))
```

Comment:
Note here the issue with PubSubClient that has a timeout of 20 seconds. So each MQTT_RECONNECT_DELAY, you could lock the program on this line for 20 seconds, meaning that it won't react to anything.
I strongly recommend a non blocking client like AsyncMqttClient

Reply:
That remark is of course important when using the wifi on esp8266 with weak signal. That's a real pain. For wired connection, you shouldn't loose your connection and PubSubClient is an easy acceptable multi arch/board solution.

Reply:
Yeah, I agree. I have the async libraries working but the problem child is OS2.3.

Both PubSubClient and UIPEthernet introduce timeouts (DNS Lookup, TCP Connection and MQTT Sync). The timeouts can be de-tuned but UIPEthernet does a lot more than just support MQTT.

The issue is really down to support for keep-alive and LWT. I'm reluctant to give up on these as "availability" is an important feature for "set and forget" devices. I'm going to play around a bit more to see if I can find a happy medium. But suspect we will need different solutions for each platform (i.e async mosquitto for OSPi, async-mqtt for OS3.0 and something cut down for OS2,3).

@jbaudoux , Let me know if you have thoughts/insight into OS2.3 solutions.

Reply:
Looking to close out this comment and appreciate any feedback.

There are currently three "timeouts" that impact MQTT network traffic and can block the main loop whilst establishing an MQTT connection as follows:

- DNS Lookup - A 15 second timeout if the network is down or the DNS server is unavailable
- TCP Connect - A 25 second timeout if the MQTT server port is closed
- MQTT Sync - A 15 second timeout if the MQTT Broker service is not responding

Note that these are not additive but rather the first timeout will exit the connection attempt so the maximum block would be 25 seconds (numbers based on OS2.3 which is worst case). This is similar for other outbound traffic such as HTTP Stations and IFTTT Notifications. However, MQTT is more susceptible to blocking as the "Keep Alive" communication happens on a very regular basis e.g. typically acks are sent every 15-60 second. Meaning that with default settings the main loop could be blocked for 25 seconds every minute if say the MQTT server was having issues.

I would like to maintain keep-alives as they provide a mechanism to detect OS going offline and allow for automations that alert the user. Very useful in "set and forget" equipment. Unfortunately, there aren't async versions of the libraries available across all platforms and for those platforms that do have libraries available (OSPi and OS3.0 WiFi) then all of the networking code base would need to be re-factored to transition to async communication. So I'm proposing is to remain with the existing synchronous architecture but adjust the KeepAlive and BackOff periods to maintain a balance between responsiveness and ensuring minimal disruption to the main event loop.

A very conservation approach would be to extend the KeepAlive period to 60 minutes (i.e. the Broker will recognise that OS is offline after an hour). To ensure the firmware doesn't block the event loop every hour for around 25 seconds then we can set a back-off period of 24 hours i.e. following an MQTT connection failure then Firmware would wait for 24 hours before attempting to reconnect. This is very conservative but would have limited impact on the main sprinkler functionality. The downside is that if there is a patchy network signal then the user may end up with significant number of "blackout" periods with no MQTT notifications.

An more aggressive approach would be to re-code the various timeouts in the libraries, say to 6 seconds. This would impact more than just MQTT functionality but would mean the KeepAlive period could be say 5 minutes and the back-off period down around an hour. It would mean modifying the libraries as they don't have apis for these settings.

@rayshobby , I would appreciate thoughts on this to get the balance right.

Reply:
I think this issue is mainly in case of wifi connection with esp8266. Using asyncTCP is for sure the best way to solve this. And once esp8266 core 3.0.0 will be released, you could normally drop UIPEthernet. But that means splitting code in arduino for esp8266 and non esp8266.